Removed client/server failover tests to separate test suites.

kobalski opened the pull request:
We have two new test classes `JettyClientFailOverTest` and `TomcatClientFailOverTest`.
Fixes #7180

serkan-ozal commented:
run-lab-run

mesutcelik commented:
:+1:  I see that is pure reorganization of fail-over test cases.

pveentjer commented on hazelcast-wm/src/test/java/com/hazelcast/wm/test/WebFilterClientFailOverTests.java:
```diff
@@ -0,0 +1,97 @@
+        CookieStore cookieStore = new BasicCookieStore();
+        assertEquals("true", executeRequest("write", serverPort1, cookieStore));
+        assertEquals("value", executeRequest("read", serverPort1, cookieStore));
+        hz.shutdown();
```
please add a whiteline after before the shutdown and the other steps. Then the layout explains you have multiple steps.

pveentjer commented on hazelcast-wm/src/test/java/com/hazelcast/wm/test/WebFilterClientFailOverTests.java:
```diff
@@ -0,0 +1,97 @@
+        assertEquals("null", executeRequest("read", serverPort1, cookieStore));
+        hz = Hazelcast.newHazelcastInstance(
+                new FileSystemXmlConfig(new File(sourceDir + "/WEB-INF/", "hazelcast.xml")));
+        Thread.sleep(8000);
```
Can't we use assertTrueEventually? Doing Thread.sleep is a testing smell in a lot of cases and we should prevent it as much as possible.

pveentjer commented on hazelcast-wm/src/test/java/com/hazelcast/wm/test/WebFilterClientFailOverTests.java:
```diff
@@ -0,0 +1,97 @@
+
+    @Test
+    public void whenClusterIsDownAtBeginning_MapSizeAfterClusterIsUp_deferredWrite() throws Exception {
+        if (!testName.equals("client - not deferred")) return;
```
This doesn't follow our coding standard. We always use brackets.

pveentjer commented on hazelcast-wm/src/test/java/com/hazelcast/wm/test/WebFilterClientFailOverTests.java:
```diff
@@ -0,0 +1,97 @@
+        Thread.sleep(8000);
+        assertEquals("value", executeRequest("read", serverPort1, cookieStore));
+        assertEquals("true", executeRequest("remove", serverPort1, cookieStore));
+        assertEquals("null", executeRequest("read", serverPort1, cookieStore));
```
Why not assertNull.

pveentjer commented on hazelcast-wm/src/test/java/com/hazelcast/wm/test/WebFilterClientFailOverTests.java:
```diff
@@ -0,0 +1,97 @@
+    public void whenClusterIsDownAtBeginning_enabledDeferredWrite() throws Exception {
+        hz.shutdown();
+        CookieStore cookieStore = new BasicCookieStore();
+        assertEquals("true", executeRequest("write", serverPort1, cookieStore));
```
I would create an assert method:

assertRequest("true", "write", serverPort1, cookieStore)

And cookie stored core be created in the setup so doesn't need to be passed as argument.

pveentjer commented on hazelcast-wm/src/test/java/com/hazelcast/wm/test/WebFilterClientFailOverTests.java:
```diff
@@ -0,0 +1,97 @@
+        IMap<String, Object> map = hz.getMap(DEFAULT_MAP_NAME);
+        assertEquals(0, map.size());
+    }
+
```
Why so many empty lines?

pveentjer commented on hazelcast-wm/src/test/java/com/hazelcast/wm/test/WebFilterClientFailOverTests.java:
```diff
@@ -0,0 +1,97 @@
+
+public abstract class WebFilterClientFailOverTests extends AbstractWebFilterTest {
+
+    String testName;
```
Why no private access modifier?

pveentjer commented on hazelcast-wm/src/test/java/com/hazelcast/wm/test/WebFilterClientFailOverTests.java:
```diff
@@ -0,0 +1,97 @@
+    @Test
+    public void whenClusterIsDown_enabledDeferredWrite() throws Exception {
+        CookieStore cookieStore = new BasicCookieStore();
+        assertEquals("true", executeRequest("write", serverPort1, cookieStore));
```
This test is doing too much. You should not do any assert in the initialization of the test. 

A bit cleaner is:
```
 @Test
    public void whenClusterIsDown_enabledDeferredWrite() throws Exception {
        CookieStore cookieStore = new BasicCookieStore();
        executeRequest("write", serverPort1, cookieStore);
        executeRequest("read", serverPort1, cookieStore);
        
	hz.shutdown();
        assertEquals("value", executeRequest("read", serverPort1, cookieStore));
        assertEquals("true", executeRequest("remove", serverPort1, cookieStore));
        assertEquals("null", executeRequest("read", serverPort1, cookieStore));
     
        hz = Hazelcast.newHazelcastInstance(
                new FileSystemXmlConfig(new File(sourceDir + "/WEB-INF/", "hazelcast.xml")));
        Thread.sleep(8000);
        assertEquals("true", executeRequest("write", serverPort1, cookieStore));
        assertEquals("value", executeRequest("read", serverPort1, cookieStore));
    }
```

And I'm still not that happy because of the assert immediately after the hz.shutdown. Are these the part of the logic to be tested or are you still setting up your system. Imho you are still setting up your system and therefor they should not be asserts. Please check the wiki:
https://hazelcast.atlassian.net/wiki/display/EN/Testing+Practices

pveentjer commented on hazelcast-wm/src/test/java/com/hazelcast/wm/test/WebFilterClientFailOverTests.java:
```diff
@@ -0,0 +1,97 @@
+    }
+
+    @Test
+    public void whenClusterIsDownAtBeginning_MapSizeAfterClusterIsUp_NonDeferredWrite() throws Exception {
```
The name of the test doesn't explain what is happening. So it isn't providing a lot of value.

serkan-ozal commented:
:+1:

bilalyasar commented:
:+1:

kobalski commented:
run-lab-run

kobalski commented:
@pveentjer Thank you for your comments. I tried to correct most of the issues, but some of them will require too much work. This PR only solves a test failure, I just moved some of the problematic tests to a new TestSuite. Most of your concerns are also present on other tests of session replication, so changes made here should also applied all of the other tests. I added a refactoring task to our technical dept PRD. I will do a refactor considering your suggestions, but not right now.
